### PR TITLE
feat(certificate): create a compat layer for provider generation

### PR DESCRIPTION
**Pull request by steeling**
The compat layer is used to convert from the old style options
to the new style MRC. This is useful so that we don't have to manage
two implementations of the certificate.Manager.

Signed-off-by: Sean Teeling <seanteeling@microsoft.com>

part of #4713

### CODE REVIEW

File: pkg/apis/config/v1alpha2/meshrootcertificate.go [hunk 1]
```diff
@@ -66,6 +66,9 @@ type VaultProviderSpec struct {
 	// Host specifies the name of the Vault server
 	Host string `json:"host"`
 
+	// Port specifies the port of the Vault server
+	Port int `json:"port"`
```
Comment (jaellio) [inline pkg/apis/config/v1alpha2/meshrootcertificate.go:70]: Also update `config_mesh_root_certificate.yaml` to include the port value

Comment (jaellio) [reply pkg/apis/config/v1alpha2/meshrootcertificate.go:70]: I was :)

File: cmd/osm-bootstrap/crds/config_mesh_root_certificate.yaml [hunk 2]
```diff
@@ -85,13 +85,17 @@ spec:
...
                         - host
+                        - port
                         - role
                         - protocol
                         - token
                       properties:
                         host:
                           description: Host name for the Vault server
                           type: string
+                        port:
```
Comment (jaellio) [inline cmd/osm-bootstrap/crds/config_mesh_root_certificate.yaml:96]: Should vault.port also be added to `values.yaml` and `values.schema.json`?

Comment (steeling) [reply cmd/osm-bootstrap/crds/config_mesh_root_certificate.yaml:96]: done!

File: cmd/osm-controller/validate.go [hunk 3]
```diff
@@ -32,18 +30,9 @@ func validateCLIParams() error {
 }
 
 func validateCertificateManagerOptions() error {
```
Comment (jaellio) [inline cmd/osm-controller/validate.go:32]: Why is this only called in the osm-controller?

File: pkg/certificate/providers/config.go [hunk 4]
```diff
@@ -29,227 +29,153 @@ const (
...
 
 // getHashiVaultOSMCertificateManager returns a certificate manager instance with Hashi Vault as the certificate provider
-func (c *Config) getHashiVaultOSMCertificateManager(options VaultOptions) (*certificate.Manager, debugger.CertificateManagerDebugger, error) {
-	if _, ok := map[string]interface{}{"http": nil, "https": nil}[options.VaultProtocol]; !ok {
-		return nil, nil, fmt.Errorf("value %s is not a valid Hashi Vault protocol", options.VaultProtocol)
+func (c *MRCProviderGenerator) getHashiVaultOSMCertificateManager(mrc *v1alpha2.MeshRootCertificate) (certificate.Issuer, error) {
+	provider := mrc.Spec.Provider.Vault
+	if _, ok := map[string]interface{}{"http": nil, "https": nil}[provider.Protocol]; !ok {
+		return nil, fmt.Errorf("value %s is not a valid Hashi Vault protocol", provider.Protocol)
 	}
```
Comment (jaellio) [inline pkg/certificate/providers/config.go:128]: Why does this need to be validated here if it is already validated in `validate()` for Hashi Vault?

File: cmd/osm-controller/validate.go (hunk 3 above)
Comment (steeling) [reply cmd/osm-controller/validate.go]: removed, it's an unnecessary additional validation

File: pkg/certificate/providers/config.go (hunk 4 above)
Comment (steeling) [reply pkg/certificate/providers/config.go]: good catch! removed

**Review by nojnhuh — APPROVED** [review]
Left a few comments for small stuff, but overall LGTM.

**Review by jaellio — APPROVED** [review]

File: pkg/certificate/providers/types.go [hunk 5]
```diff
@@ -33,31 +32,19 @@ var (
...
-	// certManagerOptions is the options for 'cert-manager.io' certiticate provider
-	certManagerOptions CertManagerOptions
-
-	msgBroker *messaging.Broker
+	AsProviderSpec() v1alpha2.ProviderSpec
 }
 
 // TresorOptions is a type that specifies 'Tresor' certificate provider options
 type TresorOptions struct {
 	// No options at the moment
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/certificate/providers/types.go:46]: Yup! basically how to get the secret is one of the "options"

File: charts/osm/values.schema.json [hunk 6]
```diff
@@ -1267,6 +1267,12 @@
                             "description": "Hashicorp Vault host/service - where Vault is installed",
                             "type": "string"
                         },
+                        "port": {
+                            "$id": "#/properties/osm/properties/vault/properties/port",
+                            "title": "Hashicorp Vault's port",
+                            "description": "Port to use to connect to vault",
+                            "type": "integer"
```
[the remark this answers is not in the document]
Comment (steeling) [reply charts/osm/values.schema.json]: done!

**Comment by steeling** [comment]
addressed all comments and squashed to single commit

**Pull request merged**